updated access to cluster/api w/o proxy post kube v1.3.2

johnt337 opened the pull request:
Not sure exactly at what version this changed, only recently noticed in v1.3.2
Assuming this was introduced with namespaces.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubernetes/kubernetes.github.io/1011)
<!-- Reviewable:end -->

johnt337 commented:
I signed the CLA under my personal account.

pwittrock commented:
@AdoHe Do you know if this was a recent change?

AdoHe commented:
IIRC, we don't update token display in recent changes, @johnt337 would you mind providing the full `kubectl config view` content, I will confirm this.

johnt337 commented:
@AdoHe here you go.

```
apiVersion: v1
clusters:
- cluster:
    certificate-authority: credentials/ca.pem
    server: https://kube-080-test.qa.somedomain.com
  name: kube-aws-kube-080-test-cluster
contexts:
- context:
    cluster: kube-aws-kube-080-test-cluster
    namespace: default
    user: kube-aws-kube-080-test-admin
  name: kube-aws-kube-080-test-context
current-context: kube-aws-kube-080-test-context
kind: Config
preferences: {}
users:
- name: kube-aws-kube-080-test-admin
  user:
    client-certificate: credentials/admin.pem
    client-key: credentials/admin-key.pem
```

janetkuo commented:
@kubernetes/docs are we planning to support versioned docs?

pwittrock commented:
cc @jaredbhatti 
@janetkuo Not in the immediate future.  Yes as a long term goal.

pwittrock commented:
@AdoHe any updates?  Would you compare the a `kubectl config` from the release-1.2 branch with the release-1.3 branch to see if you can reproduce?

AdoHe commented:
@johnt337 I test this on `1.3.1` and `1.2.5`, below is the result:
```
[root@localhost kubernetes]# kc version
Client Version: version.Info{Major:"1", Minor:"3+", GitVersion:"v1.3.1-dirty", GitCommit:"fe4aa01af2e1ce3d464e11bc465237e38dbcff27", GitTreeState:"dirty", BuildDate:"2016-08-19T01:42:55Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"3+", GitVersion:"v1.3.1-dirty", GitCommit:"fe4aa01af2e1ce3d464e11bc465237e38dbcff27", GitTreeState:"dirty", BuildDate:"2016-08-19T01:42:55Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}

[root@localhost kubernetes]# kc config view
apiVersion: v1
clusters: []
contexts:
- context:
    cluster: ""
    user: cluster-admin
  name: local
current-context: local
kind: Config
preferences: {}
users:
- name: cluster-admin
  user:
    token: testtoken
```

```
[root@localhost kubernetes]# kc version
Client Version: version.Info{Major:"1", Minor:"2", GitVersion:"v1.2.5", GitCommit:"25eb53b54e08877d3789455964b3e97bdd3f3bce", GitTreeState:"clean"}
Server Version: version.Info{Major:"1", Minor:"2", GitVersion:"v1.2.5", GitCommit:"25eb53b54e08877d3789455964b3e97bdd3f3bce", GitTreeState:"clean"}

[root@localhost kubernetes]# kc config view
apiVersion: v1
clusters: []
contexts:
- context:
    cluster: ""
    user: cluster-admin
  name: local
current-context: local
kind: Config
preferences: {}
users:
- name: cluster-admin
  user:
    token: testtoken
```

janetkuo commented:
This needs a rebase now

janetkuo commented:
> `kubectl config view` still has token here, and I checked the config struct, we didn't modify it in recent releases. would you mind recheck your environment?

@johnt337 it seems that `token` is still supported post 1.3.x. Would you verify that?

devin-donnelly commented:
@AdoHe or @janetkuo , is this good to go? I'll just admin-merge it if so.

AdoHe commented:
@johnt337 I am afraid not good to go, please see my old comment.

johnt337 commented:
@AdoHe, @janetkuo sorry for the delay. Per your request. I went back and checked to see if my config shows a token, but it does not.

Does this require the token be added to the config manually?

I updated my client to 1.3.6 to see if it makes a difference.

``` sh
 johnt  ⋯  jenkins  deployments  test-pipeline  KUBECONFIG=kubeconfig kubectl version                                                feature/kube ★ test-pipeline-0.0.2 ↑ 1 ● 1 ✚ 9 … 6 ⚑ 1  00:39
Client Version: version.Info{Major:"1", Minor:"3", GitVersion:"v1.3.6+ae4550c", GitCommit:"ae4550cc9c89a593bcda6678df201db1b208133b", GitTreeState:"not a git tree", BuildDate:"2016-08-30T15:45:51Z", GoVersion:"go1.7", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"3", GitVersion:"v1.3.2+coreos.0", GitCommit:"52a0d5141b1c1e7449189bb0be3374d610eb98e0", GitTreeState:"clean", BuildDate:"2016-07-19T17:45:13Z", GoVersion:"go1.6.2", Compiler:"gc", Platform:"linux/amd64"}
```

I also created three different clusters using kube-aws, kops, and minikube, but I still don't see it in the output of `kubectl config view` for any of the clusters.

``` sh
 johnt  ⋯  jenkins  deployments  test-pipeline  kubectl config view                                                                  feature/kube ★ test-pipeline-0.0.2 ↑ 1 ● 1 ✚ 9 … 6 ⚑ 1  00:42
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: REDACTED
    server: https://api.jt-alpha.somedomain.com
  name: jt-alpha.somedomain.com
- cluster:
    certificate-authority: /Users/johnt/gitroot/containers/dockerfiles/jenkins/deployments/test-pipeline/credentials/ca.pem
    server: https://test-pipeline.qa.somedomain.com
  name: kube-aws-test-pipeline-cluster
- cluster:
    certificate-authority: /Users/johnt/.minikube/ca.crt
    server: https://192.168.64.3:8443
  name: minikube
contexts:
- context:
    cluster: jt-alpha.somedomain.com
    user: jt-alpha.somedomain.com
  name: jt-alpha.somedomain.com
- context:
    cluster: kube-aws-test-pipeline-cluster
    namespace: default
    user: kube-aws-test-pipeline-admin
  name: kube-aws-test-pipeline-cluster
- context:
    cluster: minikube
    user: minikube
  name: minikube
current-context: jt-alpha.somedomain.com
kind: Config
preferences: {}
users:
- name: jt-alpha.somedomain.com
  user:
    client-certificate-data: REDACTED
    client-key-data: REDACTED
    password: XXXXXXXXXX
    username: admin
- name: jt-alpha.somedomain.com-basic-auth
  user:
    password: XXXXXXXXXX
    username: admin
- name: kube-aws-test-pipeline-admin
  user:
    client-certificate: /Users/johnt/gitroot/containers/dockerfiles/jenkins/deployments/test-pipeline/credentials/admin.pem
    client-key: /Users/johnt/gitroot/containers/dockerfiles/jenkins/deployments/test-pipeline/credentials/admin-key.pem
- name: minikube
  user:
    client-certificate: /Users/johnt/.minikube/apiserver.crt
    client-key: /Users/johnt/.minikube/apiserver.key
```

I tried to override the config with one in my current directory where the config is as follows:

``` sh
 johnt  ⋯  jenkins  deployments  test-pipeline  cat kubeconfig                                                                       feature/kube ★ test-pipeline-0.0.2 ↑ 1 ● 1 ✚ 9 … 6 ⚑ 1  00:55
apiVersion: v1
kind: Config
clusters:
- cluster:
    certificate-authority: credentials/ca.pem
    server: https://test-pipeline.qa.somedomain.com
  name: kube-aws-test-pipeline-cluster
contexts:
- context:
    cluster: kube-aws-test-pipeline-cluster
    namespace: default
    user: kube-aws-test-pipeline-admin
  name: kube-aws-test-pipeline-context
users:
- name: kube-aws-test-pipeline-admin
  user:
    client-certificate: credentials/admin.pem
    client-key: credentials/admin-key.pem
current-context: kube-aws-test-pipeline-context
```

then ran it two different ways, one with the env var `KUBECONFIG=./kubeconfig` and one with the `--kubeconfig=./kubeconfig` option passed.

Below is the output from each run.

``` sh
 johnt  ⋯  jenkins  deployments  test-pipeline  KUBECONFIG=./kubeconfig kubectl config view                                          feature/kube ★ test-pipeline-0.0.2 ↑ 1 ● 1 ✚ 9 … 6 ⚑ 1  00:55
apiVersion: v1
clusters:
- cluster:
    certificate-authority: credentials/ca.pem
    server: https://test-pipeline.qa.somedomain.com
  name: kube-aws-test-pipeline-cluster
contexts:
- context:
    cluster: kube-aws-test-pipeline-cluster
    namespace: default
    user: kube-aws-test-pipeline-admin
  name: kube-aws-test-pipeline-context
current-context: kube-aws-test-pipeline-context
kind: Config
preferences: {}
users:
- name: kube-aws-test-pipeline-admin
  user:
    client-certificate: credentials/admin.pem
    client-key: credentials/admin-key.pem
 johnt  ⋯  jenkins  deployments  test-pipeline  kubectl config view --kubeconfig=./kubeconfig                                        feature/kube ★ test-pipeline-0.0.2 ↑ 1 ● 1 ✚ 9 … 6 ⚑ 1  00:58
apiVersion: v1
clusters:
- cluster:
    certificate-authority: credentials/ca.pem
    server: https://test-pipeline.qa.somedomain.com
  name: kube-aws-test-pipeline-cluster
contexts:
- context:
    cluster: kube-aws-test-pipeline-cluster
    namespace: default
    user: kube-aws-test-pipeline-admin
  name: kube-aws-test-pipeline-context
current-context: kube-aws-test-pipeline-context
kind: Config
preferences: {}
users:
- name: kube-aws-test-pipeline-admin
  user:
    client-certificate: credentials/admin.pem
    client-key: credentials/admin-key.pem
 johnt  ⋯  jenkins  deployments  test-pipeline 
```
Anything else you want me to try?

johnt337 commented:
Also did you want lines 72 & 88 removed?
Let me know if you want me to close the PR if you think this is something related to my local config.

johnt337 commented:
I also tried to use `kubectl config use-context ...` but still don't see a token.

``` sh
 johnt  ⋯  jenkins  deployments  test-pipeline  kubectl config use-context kube-aws-test-pipeline-cluster                     0  1   feature/kube ★ test-pipeline-0.0.2 ↑ 1 ● 1 ✚ 9 … 6 ⚑ 1  01:17
switched to context "kube-aws-test-pipeline-cluster".
 johnt  ⋯  jenkins  deployments  test-pipeline  kubectl config view 2>&1 | grep -i current                                  0  1   feature/kube ★ test-pipeline-0.0.2 ↑ 1 ● 1 ✚ 9 … 6 ⚑ 1  01:34
current-context: kube-aws-test-pipeline-cluster
 johnt  ⋯  jenkins  deployments  test-pipeline  kubectl config view 2>&1 | grep -i token                                             feature/kube ★ test-pipeline-0.0.2 ↑ 1 ● 1 ✚ 9 … 6 ⚑ 1  01:35
 johnt  ⋯  jenkins  deployments  test-pipeline  kubectl config use-context jt-alpha.somedomain.com                          0  1   feature/kube ★ test-pipeline-0.0.2 ↑ 1 ● 1 ✚ 9 … 6 ⚑ 1  01:35
switched to context "jt-alpha.somedomain.com".
 johnt  ⋯  jenkins  deployments  test-pipeline  kubectl config view 2>&1 | grep -i current                                           feature/kube ★ test-pipeline-0.0.2 ↑ 1 ● 1 ✚ 9 … 6 ⚑ 1  01:35
current-context: jt-alpha.somedomain.com
 johnt  ⋯  jenkins  deployments  test-pipeline  kubectl config view 2>&1 | grep -i token                                             feature/kube ★ test-pipeline-0.0.2 ↑ 1 ● 1 ✚ 9 … 6 ⚑ 1  01:35
 johnt  ⋯  jenkins  deployments  test-pipeline                                                                              0  1   feature/kube ★ test-pipeline-0.0.2 ↑ 1 ● 1 ✚ 9 … 6 ⚑ 1  01:35
```

AdoHe commented:
@johnt337 from your `kubectl config view` output, I can see you are using certificates&bearer token&basic auth, basically bearer token&basic auth are mutually exclusive. I am not sure whether kops and minikube use token. I have checked the code, we did not modify config struct recently.

johnt337 commented:
@AdoHe so are you saying a token generated with `$(kubectl describe secret $(kubectl get secrets | grep default | cut -f1 -d ' ') | grep -E '^token' | cut -f2 -d':' | tr -d '\t')` is some how different than the one that would be generated via `$(kubectl config view | grep token | cut -f 2 -d ":" | tr -d " ")`?

Should I expected the documented way of doing this work provided I built this cluster manually?

I can connect to the clusters in all scenarios using any of the auth methods that are configured, its just the solicitation of the token per the docs returns nothing.

Given I have a way to generate tokens regardless of how the cluster was built, i'm fine with closing this PR.

Please advise.

devin-donnelly commented on docs/user-guide/accessing-the-cluster.md:
```diff
@@ -69,7 +69,7 @@ $ curl http://localhost:8080/api/
-#### Without kubectl proxy
+#### Without kubectl proxy (before v1.3.x)
 
 It is also possible to avoid using kubectl proxy by passing an authentication token
```
Remove "also"; unnecessary.

devin-donnelly commented on docs/user-guide/accessing-the-cluster.md:
```diff
@@ -85,7 +85,29 @@ $ curl $APISERVER/api --header "Authorization: Bearer $TOKEN" --insecure
-The above example uses the `--insecure` flag.  This leaves it subject to MITM
+#### Without kubectl proxy (post v1.3.x)
+
+In recent versions of Kubernetes, `kubectl config view` no longer displays the token. Use `kubectl describe secret...` to get the token for the default service account, like this:
```
"In recent versions" -> "In Kubernetes version 1.3 or later,"

devin-donnelly commented:
Couple of quick docs comments, maybe a rebase, and we're good to go.

johnt337 commented:
@devin-donnelly will apply those changes and rebase.

johnt337 commented:
I've applied the changes you requested.

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1011#-:-KReKV4_1ajiTyIIupBy:b-dawqsi)*
<!-- Sent from Reviewable.io -->

johnt337 commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[docs/user-guide/accessing-the-cluster.md, line 74 at r1](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1011#-KPPDu0h_xLvk96Cj2Ve-r1-74:-KReKpyuyxh34zFTms3t:bss4a2b) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c558fb57017ff244ef5e3a6578a1246e9ce63eb1/docs/user-guide/accessing-the-cluster.md#L74)):*
<details><summary><i>Previously, devin-donnelly wrote…</i></summary>
> Remove "also"; unnecessary.
</details>
Removed.

---

*[docs/user-guide/accessing-the-cluster.md, line 90 at r1](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1011#-KPPDu0h_xLvk96Cj2Ve-r1-90:-KReKuO06BpH1OsAOftj:b-bw8e14) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c558fb57017ff244ef5e3a6578a1246e9ce63eb1/docs/user-guide/accessing-the-cluster.md#L90)):*
<details><summary><i>Previously, devin-donnelly wrote…</i></summary>
> "In recent versions" -> "In Kubernetes version 1.3 or later,"
</details>
Updated.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1011)*
<!-- Sent from Reviewable.io -->

devin-donnelly commented:
This looks good. I'll merge.
